### ISSUE TITLE
Difficulty with model convergence for VGG16 on CIFAR-10

### ISSUE THREAD

**Issue by GoldenPotato137** [ucbrise/piranha#19]
Hi Piranha authors,

We are currently working with your repository to replicate the VGG16 on CIFAR-10 training experiment described in the associated paper.
We are using the model configuration provided in the repository: `files/models/vgg16-cifar10-norelu.json`. Our run configuration is as follows:
```json
{
    "_desc_num_parties": "Number of parties involved in the MPC computation. Should match compiled application protocol.",
    "num_parties": 3,

    "_desc_party_ips": "IP addresses of each party, ordered by party number",
    "party_ips": ["172.16.30.16", "172.16.30.7", "172.16.30.13"],

    "_desc_party_users": "Usernames for each party, ordered by party number. Can be used by scripts to SSH into machines if necessary to start piranha processes",
    "party_users": ["ubuntu"],

    "_desc_run_unit_tests": "Run unit tests",
    "run_unit_tests": false,

    "_desc_unit_test_only": "Only run unit tests, do not try to run any full Piranha applications",
    "unit_test_only": false,

    "_desc_debug_print": "Show debug output",
    "debug_print": false,

    "_desc_debug_overflow": "Test for overflow and print related debug output. Breaks security by revealing intermediary values.",
    "debug_overflow": false,

    "_desc_debug_sqrt": "Test sqrt for invalid input and print related debug output. Breaks security by revealing intermediary values.",
    "debug_sqrt": false,

    "_desc_subnet_detailed_log": "Show detailed subnet forward/backward progress debug output",
    "subnet_detailed_log": false,

    "_desc_run_name": "Descriptive name for run, used to name log files with something useful",
    "run_name": "hello_world_run",

    "_desc_network": "Path to NN architecture to use for the run",
    "network": "files/models/vgg16-cifar10-norelu.json",

    "_desc_custom_epochs": "Enable custom number of epochs. Otherwise set by size of learning rate schedule",
    "custom_epochs": false,

    "_desc_custom_epoch_count": "Number of epochs to train for, if custom_epochs is set",
    "custom_epoch_count": 3,

    "_desc_custom_iterations": "Enable custom number of iterations. Otherwise set by training dataset size.",
    "custom_iterations": false,

    "_desc_custom_iteration_count": "Number of iterators to train per epoch, if custom_iterations is set",
    "custom_iteration_count": 0,

    "_desc_custom_batch_size": "Enable custom batch size. Otherwise set by architecture configuration.",
    "custom_batch_size": true,

    "_desc_custom_batch_size_count": "Desired custom batch size",
    "custom_batch_size_count": 32,

    "_desc_nn_seed": "Seed for NN initialization",
    "nn_seed": 343934585,

    "_desc_preload": "Preload weights from a snapshot directory instead of training from scratch",
    "preload": false,

    "_desc_preload_path": "Directory path from which to preload network weights",
    "preload_path": "",

    "_desc_lr_schedule": "Learning rate schedule, in negative powers of 2 (e.g. 3 -> learning rate of 2^-3). Assumes that the number of LR exponents matches the desired number of training epochs",
    "lr_schedule": [3, 3, 3, 4, 5, 5, 6, 7, 8, 9],

    "_desc_test_only": "Only run NN test, skip training (useful if weights have been preloaded)",
    "test_only": false,

    "_desc_inference_only": "Only run inference (forward pass), not backward pass training",
    "inference_only": false,

    "_desc_no_test": "Do not run testing after training epochs",
    "no_test": true,

    "_desc_last_test": "Only run a test pass after the last training epoch",
    "last_test": false,

    "_desc_iteration_snapshots": "Take snapshots at each training iteration",
    "iteration_snapshots": false,

    "_desc_test_iteration_snapshots": "Take snapshots of a '1PC' test network running the same data",
    "test_iteration_snapshots": false,

    "_desc_epoch_snapshots": "Take snapshots after each training epoch",
    "epoch_snapshots": false,

    "_desc_eval_accuracy": "Evaluation: print training/test accuracy",
    "eval_accuracy": true,

    "_desc_eval_inference_stats": "Evaluation: print runtime and communication statistics for each inference forward pass",
    "eval_inference_stats": false,

    "_desc_eval_train_stats": "Evaluation: print runtime and communication statistics for each training forward-backward pass",
    "eval_train_stats": false,

    "_desc_eval_fw_peak_memory": "Evaluation: print peak memory usage during each forward pass",
    "eval_fw_peak_memory": false,

    "_desc_eval_bw_peak_memory": "Evaluation: print peak memory usage during each backward pass",
    "eval_bw_peak_memory": false,
    
    "_desc_eval_epoch_stats": "Evaluation: print cumulative runtime and communication statistics for each training epoch",
    "eval_epoch_stats": true,

    "_desc_print_activations": "Print output activations for each layer every forward pass",
    "print_activations": false,

    "_desc_print_deltas": "Print input gradient to each layer every backward pass",
    "print_deltas": false,
    
    "_desc_debug_all_forward": "Print debug information for all layer forward passes",
    "debug_all_forward": false,

    "_desc_debug_all_backward": "Print debug information for all layer backward passes",
    "debug_all_backward": false
}
```
However, we are finding that the model struggles to converge. After running for 10 epochs, the test accuracy is only around 13%, which is barely better than random guessing (10% for CIFAR-10).

To investigate further, we made a small modification to the code to display a moving average of the cross-entropy loss (window size = 20). You can see our changes in this fork: https://github.com/GoldenPotato137/piranha. We observed that the loss remains consistently high, fluctuating around 2.25, which aligns with the poor accuracy results.

We have attempted to troubleshoot this by modifying the learning rate schedule and even adding Layer Normalization layers to the architecture, but unfortunately, none of these changes have resolved the convergence issue.

We suspect we might be making a mistake in our setup or configuration. Could you please provide some guidance on what we might be doing wrong?

Any help you could offer would be greatly appreciated. Thank you for your time and consideration.
